core/src/banking_stage.rs

```diff
@@ -378,13 +383,257 @@ pub enum ForwardOption {
+                                        for processed_packet in &packets_to_process[start..end] {
+                                            buffered_packet_batches
+                                                .message_hash_to_transaction
+                                                .remove(processed_packet.message_hash());
+                                        }
+                                    },
+                                );
+                                retryable_transaction_indexes
+                                    .iter()
+                                    .map(|i| packets_to_process[*i].clone())
+                                    .collect_vec()
+                            } else {
+                                packets_to_process
+                            }
+                        })
+                        .collect::<MinMaxHeap<_>>();
+
+                buffered_packet_batches.packet_priority_queue = retryable_packets;
+
+                // Assert unprocessed queue is still consistent
+                assert_eq!(
+                    buffered_packet_batches.packet_priority_queue.len(),
+                    buffered_packet_batches.message_hash_to_transaction.len()
+                );
+            }
+            Self::VoteStorage(latest_unprocessed_votes, VoteSource::Tpu) => {
+                // Based on the stake distribution present in the supplied bank, drain the unprocessed votes
+                // from each validator using a weighted random ordering. Votes from validators with
+                // 0 stake are ignored.
+                // TODO: Add proper error handling in case bank is missing?
+                let bank = bank.unwrap();
+
+                let latest_votes_per_pubkey = latest_unprocessed_votes
+                    .latest_votes_per_pubkey
+                    .read()
+                    .unwrap();
+
+                let retryable_votes = weighted_random_order_by_stake(&bank)
+                    .filter_map(|pubkey| {
+                        if let Some(lock) = latest_votes_per_pubkey.get(&pubkey) {
+                            if let Ok(latest_vote) = lock.write() {
+                                if let Ok(mut latest_vote) = latest_vote.try_borrow_mut() {
+                                    if !latest_vote.is_processed() {
+                                        let packet = latest_vote.clone();
+                                        latest_vote.clear();
+                                        latest_unprocessed_votes
+                                            .size
+                                            .fetch_sub(1, Ordering::AcqRel);
+                                        return Some(packet);
+                                    }
+                                }
+                            }
+                        }
+                        None
+                    })
+                    .chunks(batch_size)
+                    .into_iter()
```

Comment:
Can we reorder this so that we immediately apply the `.map(&DeserializedVotePacket::get_vote_packet)` to the chunks, that way we clone as little as possible in `collect_vec()` later

Reply:
Not sure exactly what you mean here, the `get_vote_packet` just creates a shared pointer to the packet. We still need to keep track of the vote packets to insert the retryable ones back in.